### ISSUE TITLE
Initialization mechanism for DicomService

### ISSUE THREAD

**Issue by gofal** [fo-dicom/fo-dicom#650]
### Expected behavior
In class DicomService there should be a virtual Method "Initialize" or a event "OnInitialize" to give the possibility to do save initialization before any of the the asynchron On..Request happen (eg. create a database connection)

### Actual behavior
The dicomserver (or the tcp-listener) runs in thread1. when an association commes, it creates an instance of StoreSCP. the constructor of StoreSCP first calls the base constructor of DicomServerice. this base constructor creates an new task in "PduListener = ListenAndProcessPDUAsync();" and within this new thread2 the streams are processed and the calls into OnReceiveAssociationRequest and so on are called. while this runs in thread2 the constructor of StoreSCP continues in thread1. if you do some long initialisation in constructor it might happen that the console.writeline in OnReceiveAssociationRequest happens before the console.writeline of the constructor. To solve this you need some synchronization mechanism in these methods.
For example the base constructor of DicomService could call the virtual method "Initialize" before starting the thread2 with "ListenAndProcessPduAsync"

### Steps to reproduce the behavior
Add some long lasting initialization in costructor of a DicomService derived class. the method OnReceiveAssociationRequest will happen before the initialization is ready

**Comment by gofal** [comment]
I had observed this with the latest released version 3.0.2.
In the current development branch there is already a solution with commit "Improvements in asynchronous networking API. Connected to #331 Connected to #525 Connected to #574 (#584)" on 24.08.2017 by Anders Gustafsson.
Now the ListenAndProcessPduAsync is started after the constructor of the super class has been executed.
So this issue is obsolete.

**Comment by a21862** [comment]
@gofal 
Would you mind having a look at a C-Store SCP issue ( desktop v4.0.3 ) I recently try to resolve . I think it may relate to this issue.  Thanks.

Actually, I do not do any lasting initialization in costructor of a DicomService.  However, each time a new C-Store SCU tries to send Dicom File to the C-Store SCP, the first attempt would definitely fail, but the subsequent operations work correctly. 

The log of the first attempt:

T: 2019/10/12 10:01:19]	M: Info::{callingAE} <- Association request:{association}     
T: 2019/10/12 10:01:19]	M: Info::{logId} -> Association accept:{association}        
T: 2019/10/12 10:01:20]	M: Info::{logId} <- {dicomMessage}
T: 2019/10/12 10:01:20]	M: Error::Exception processing P-Data-TF PDU: {@error}        
T: 2019/10/12 10:01:20]	M: Error::Error during close attempt: {@error}

The log of the subsequent operations

T: 2019/10/12 10:02:32]	M: Info::{callingAE} <- Association request:{association}                                        
T: 2019/10/12 10:02:32]	M: Info::{logId} -> Association accept:{association}                    
T: 2019/10/12 10:02:32]	M: Info::{logId} <- {dicomMessage}             
T: 2019/10/12 10:02:32]	M: Info::{logId} -> {dicomMessage}                      
T: 2019/10/12 10:02:32]	M: Info::{logId} <- Association release request               
T: 2019/10/12 10:02:32]	M: Info::{logId} -> Association release response             
T: 2019/10/12 10:02:32]	M: Error::Error during close attempt: {@error}
T: 2019/10/12 10:02:32]	M: Error::Error during close attempt: {@error}         

In addition, I notice that there is always a "M: Error::Error during close attempt: {@error} " in the end of each operation no matter it fails or not. I have no clue of it. 

My codes follow the C-Store SCP project of the Samples.Desktop examples. There are no major changes on top of the example codes.
